### PR TITLE
Add source_model() in adapter models

### PR DESCRIPTION
**Pull request by Guiguiprim**
As a mean of getting the inner model, useful to be able to push into the underlying model. (This already exists in the CPP API.)

Also implement `as_any()` on adapters model otherwise the `source_model()` is a bit useless.

### CODE REVIEW

File: internal/core/model/adapters.rs [hunk 1]
```diff
@@ -1065,6 +1067,11 @@ where
         container.wrapped_model.model_tracker().attach_peer(container.as_ref().model_peer());
         Self(container)
     }
+
+    /// Returns a reference to the inner model
+    pub fn source_model(&self) -> Pin<&M> {
```
Comment (tronical) [inline internal/core/model/adapters.rs:1072]: Interesting, is there a particular reason for pinning here? If, like in your test, the source model is an `Rc<dyn Model>` (aka `ModelRc`) then wouldn't a `&M` be fine as well?

**Comment by ogoffart** [comment]
Thanks for the patch.
I also wonder why the Pin.

I'm thinking that we probbly should add source_model to all our adapter for consistancy.

File: internal/core/model/adapters.rs (hunk 1 above)
Comment (Guiguiprim) [reply internal/core/model/adapters.rs:1072]: We can't return `&M` as trying to do so would result in a compile error `return a reference to a temporary` (a.k.a the `Pin<>` from the call to `ModelChangeListenerContainer::get()`)

Comment (Guiguiprim) [reply internal/core/model/adapters.rs:1072]: Or maybe with some unpining ?

Comment (Guiguiprim) [reply internal/core/model/adapters.rs:1072]: I can unpin, and I should !

Comment (Guiguiprim) [reply internal/core/model/adapters.rs]: Fixed

**Review by ogoffart — APPROVED** [review]

**Review by ogoffart — APPROVED** [review]

**Comment by ogoffart** [comment]
Thanks!!

**Pull request merged**